pageserver/src/walingest.rs

```diff
@@ -400,39 +398,72 @@ impl WalIngest {
                 // we can't validate the remaining number of bytes without parsing
                 // the tuple data.
-                if (xlrec.flags & pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED) != 0 {
-                    timeline.put_wal_record(
-                        lsn,
-                        RelishTag::Relation(RelTag {
-                            forknum: pg_constants::VISIBILITYMAP_FORKNUM,
-                            spcnode: decoded.blocks[0].rnode_spcnode,
-                            dbnode: decoded.blocks[0].rnode_dbnode,
-                            relnode: decoded.blocks[0].rnode_relnode,
-                        }),
-                        decoded.blocks[0].blkno / pg_constants::HEAPBLOCKS_PER_PAGE as u32,
-                        ZenithWalRecord::ClearVisibilityMapFlags {
-                            heap_blkno: decoded.blocks[0].blkno,
-                            flags: pg_constants::VISIBILITYMAP_VALID_BITS,
-                        },
-                    )?;
-                }
-                if (xlrec.flags & pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED) != 0
-                    && decoded.blocks.len() > 1
+                if (xlrec.flags & pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED) != 0
+                    || ((xlrec.flags & pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED) != 0
+                        && decoded.blocks.len() > 1)
                 {
-                    timeline.put_wal_record(
-                        lsn,
-                        RelishTag::Relation(RelTag {
-                            forknum: pg_constants::VISIBILITYMAP_FORKNUM,
-                            spcnode: decoded.blocks[1].rnode_spcnode,
-                            dbnode: decoded.blocks[1].rnode_dbnode,
-                            relnode: decoded.blocks[1].rnode_relnode,
-                        }),
-                        decoded.blocks[1].blkno / pg_constants::HEAPBLOCKS_PER_PAGE as u32,
-                        ZenithWalRecord::ClearVisibilityMapFlags {
-                            heap_blkno: decoded.blocks[1].blkno,
-                            flags: pg_constants::VISIBILITYMAP_VALID_BITS,
-                        },
-                    )?;
+                    if (xlrec.flags
+                        & (pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED
+                            | pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED))
+                        == (pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED
+                            | pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED)
+                        && decoded.blocks.len() > 1
+                        && pg_constants::HEAPBLK_TO_MAPBLOCK(decoded.blocks[0].blkno)
+                            == pg_constants::HEAPBLK_TO_MAPBLOCK(decoded.blocks[1].blkno)
+                    {
+                        timeline.put_wal_record(
+                            lsn,
+                            RelishTag::Relation(RelTag {
```

Comment:
These put_wal_record() calls are very repetetive. Maybe construct the RelishTag before these if-statements into a local variable, to remove some of the repetition and make the code shorter.

Reply:
I added get_vm_block_tag function.
I prefer to use function rather than store tag in local variable because this tag is actually rarely needed, since only small percent of heap WAL records update VM.